### PR TITLE
Add json convertor from user request

### PR DESCRIPTION
**Pull request by shiranr**
Add json converter from user request to an object.
Make sure we validate input
Add tests
Add sort and filter for the list of Analyzer Results according to the README logic.

### CODE REVIEW

**Review by SharonHart — CHANGES_REQUESTED** [review]

File: presidio-anonymizer/app.py [hunk 1]
```diff
@@ -1,21 +1,27 @@
...
+    content = request.get_json()
+    if not content:
+        return "Invalid request json", 400
+    try:
+        data = AnonymizerEngineRequest(content)
+    except InvalidParamException as e:
+        return e.err, 400
+    except Exception:
+        # TODO add logger 2652
+        return "Internal server error", 500
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:21]: Since this is an API it is a bad practice to return internal errors.
It is ok if the user is using the lib itself (the engine). 
So it is better to mask them with 500 internal error and log them on the server side.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        :return: AnalyzerResult
+        """
+        cls.__validate_fields(content)
+        return cls(content.get("entity_type"), content.get("score"),
+                   content.get("start"),
+                   content.get("end"))
+
+    @classmethod
+    def __validate_fields(cls, content):
+        for field in ["start", "end", "score", "entity_type"]:
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:98]: Very nice comment and prettifying of the code.
But I do want the specific field remark, I think it is better for the user. :)

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                # not the same object
+                if not result_a.__eq__(other):
+                    # if there is an object containing result_a
+                    # or an object with equal indices and higher score pass
+                    if other.contains(result_a) or \
+                            (result_a.equal_indices(
+                                other) and result_a.score < other.score):
+                        break
+                if index == len(self) - 1:
+                    obj_set.add(result_a)
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:62]: I went over it with Nava and Itye this is a complex logic, I know. please read the method documentation.
the loop suggested doesn't solve the case we need :)

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    def get_transformation(self, analyzer_result: AnalyzerResult):
+        """
+        Get the right transformation from the list.
+
+        When transformation does not exist, we fall back to default.
+        :param analyzer_result: the result we are going to do the transformation on
+        :return: transformation
+        """
+        transformation = self._transformations.get(analyzer_result.entity_type)
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:48]: it is python get :) it's a class helper method.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        if results is None or len(results) == 0:
+            raise InvalidParamException("Invalid input, "
+                                        "analyzer results can not be empty")
+        for result in results:
+            self._analysis_results.append(
+                AnalyzerResult.validate_and_create(result))
+
+    def __handle_transformations(self, data):
+        transformations = data.get("transformations")
+        if transformations is not None:
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:80]: No. We use default.

File: presidio-anonymizer/app.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SharonHart) [reply presidio-anonymizer/app.py:21]: 1. Can it be done in a global exception handler then? what if we add another endpoint and someone forgets to add this? 
2. Below, you call 'text = AnonymizerEngine().anonymize(data)'. If it will fail, you aren't masking it right?

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            for key, transformation in transformations.items():
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:81]: If I understand correctly, I rather not change the map we are running over inside of the loop.

[the remark this answers is not in the document]
Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:81]: Yes, it will also help readability IMO

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:48]: What do you mean? _transformations is a dictionary?
self._transformations.get(analyzer_result.entity_type, self._transformations.get("default")) won't work?

File: presidio-anonymizer/app.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:21]: sure.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:48]: _transformations is a dictionary. yes. 
self._transformations.get(...)
The get is python dictionary.
analyzer result is an object which contains entity_type.
This is more complicated: 
self._transformations.get(analyzer_result.entity_type, self._transformations.get("default")) won't work?

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                anonymizer = self.__get_anonymizer(transformation)
+                transformation["anonymizer"] = anonymizer
+                self._transformations[key] = transformation
+
+    def __handle_text(self, data):
+        self._text = data.get("text")
+        if self._text == "" or self._text is None:
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:88]: DONE.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            raise InvalidParamException("Invalid input, text can not be empty")
+
+    def __get_anonymizer(self, transformation):
+        anonymizer_type = transformation.get("type").lower()
+        anonymizer_class = self.anonymizers.get(anonymizer_type)
+        if anonymizer_class is None:
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:94]: DONE.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:62]: WOW!!!!!!! REALLY REALLY NICE!
You would get an applause GIF if it was possible.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:81]: for key, transformation in transformations.items():
    transformation["anonymizer"] = get_anonymizer(transformation)
    transformations[key] = transformation

VS

transformations = {k: append_anonymizer(v) for k, v in transformations.items()}

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    def __eq__(self, other):
+        """
+        Check two results are equal by using all class fields.
+
+        :param other: another analyzer_result
+        :return: bool
+        """
+        return self.equal_indices(
+            other) and self.entity_type.__eq__(
+            other.entity_type) and self.score is other.score
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:72]: I think that there is an issue comparing floats this way in python:
https://stackoverflow.com/questions/5595425/what-is-the-best-way-to-compare-floats-for-almost-equality-in-python

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+    def __init__(self,
+                 data: dict):
+        """Handle text replacement for PIIs with requested transformations.
+
+        :param data: a map which contains the transformations, analyzer_results and text
+        """
+        self._transformations = {}
+        self._analysis_results = AnalyzerResults()
+        self.__validate_and_insert_input(data)
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:34]: Why do you need the double underscore here? Do you plan to have another colliding method with this name?

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        return self._analysis_results
+
+    def __validate_and_insert_input(self, data: dict):
+        self.__handle_analyzer_results(data)
+        self.__handle_text(data)
+        self.__handle_transformations(data)
+
+    def __handle_analyzer_results(self, data):
+        results = data.get("analyzer_results")
+        if results is None or len(results) == 0:
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:71]: if results: (not sure if already commented on it, don't see it cahnged in the other PR as well)

File: presidio-anonymizer/app.py [hunk 12]
```diff
@@ -1,21 +1,27 @@
-"""TODO: Flask application instantiation and route definition."""
-
 from flask import Flask, request
 
+from presidio_anonymizer.anonymizer_engine import AnonymizerEngine
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:3]: I did but for some reason when I run pytest it doesn't recognize it.

File: presidio-anonymizer/app.py [hunk 13]
```diff
@@ -1,21 +1,27 @@
...
+from presidio_anonymizer.entities.engine_request import AnonymizerEngineRequest
+from presidio_anonymizer.entities.invalid_exception import InvalidParamException
+
 app = Flask(__name__)
 
 
-# TODO add method for get supported anonymizers. 2629
 @app.route("/anonymize", methods=["POST"])
 def anonymize():
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:11]: Already done. :)

File: presidio-anonymizer/app.py [hunk 14]
```diff
@@ -1,21 +1,27 @@
...
+        return e.err, 400
+    except Exception:
+        # TODO add logger 2652
+        return "Internal server error", 500
+    text = AnonymizerEngine().anonymize(data)
+    return text
 
 
-if __name__ == "__main__":
+if __name__ == "main":
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:26]: Done.

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 15]
```diff
@@ -12,30 +12,17 @@ class AnonymizerEngine:
...
-        :param analyze_results: The results of the analyzer of PIIs locations and scores
+        :param data: a map which contains the transformations, analyzer_results and text
         """
-        self.transformations = transformations
-        self.analyze_results = analyze_results
-        self.text = text
-        self._end_point = len(text)
 
-    def anonymize(self):
+    def anonymize(self, engine_request: AnonymizerEngineRequest):
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:25]: Done

File: presidio-anonymizer/presidio_anonymizer/anonymizers/fpe.py [hunk 16]
```diff
@@ -3,7 +3,7 @@
 
 Uses FF1 algorithm for the encryption.
 """
-from anonymizers.anonymizer import Anonymizer
+from presidio_anonymizer.anonymizers.anonymizer import Anonymizer
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/fpe.py:6]: lets talk about this :)

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+            raise InvalidParamException("Invalid input, "
+                                        "analyzer results can not be empty")
+        for result in results:
+            self._analysis_results.append(
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:75]: you append to an empty list right?
self._analysis_results = [AnalyzerResult.validate_and_create(result) for result in results]

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
+"""
+AnalyzerResult is the exact copy of the recognizer result.
+
+Represents the findings of detected entity.
+"""
+from presidio_anonymizer.entities.invalid_exception import InvalidParamException
+
+
+class AnalyzerResult(object):
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:9]: Done

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:98]: Done

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    - One PII - uses a given or default transformation to anonymize and replace the PII
+    text entity.
+    - Full overlap of PIIs - When one text have several PIIs, the PII with the higher
+    score will be taken.
+    Between PIIs with similar scores, the selection will be arbitrary.
+    - One PII is contained in another - anonymizer will use the PII with larger text.
+    - Partial intersection - both will be returned concatenated.
+    """
+
+    def to_sorted_set(self, reverse=False):
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:20]: Done.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+        """
+        Manipulate the list.
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:22]: Done.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+        Only insert results which are:
+        1. Indices are not contained in other result.
+        2. Have the same indices as other results but with larger score.
+        :return: set
+        """
+        obj_set = set()
+        for result_a in self:
+            for index in range(len(self)):
+                other = self.__getitem__(index)
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:52]: Not relevant anymore but thanks :)

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+                # not the same object
+                if not result_a.__eq__(other):
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:54]: done.

**Review by SharonHart — APPROVED** [review]
Just another two minor comments.
And, I think the some of the classes deserve.a true unit test, especially the ResultList one.
We are actually not writing true unit tests, I think that in analyzer as well, but in some cases it very beneficial and easy to get all edge-cases.

File: presidio-anonymizer/app.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:3]: Done.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 17 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:75]: yes I am but I think the other way is more readable :)

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 11 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:71]: no, you commented on something else :)

Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:71]: Done.

**Comment by omri374** [comment]
@shiranr can we close this PR?
